### DISCUSSION TITLE
Upcoming version 1.0.0 release

### DISCUSSION THREAD

**Discussion by 444B** [444B/streamlit-analytics2#71] [Announcements]
Hi all
I wanted to accounce that we are driving to a version 1.0.0 release for streamlit-analytics2
Since we use semantic versioning, this does mean that that there are some breaking changes expected.

So, what could be worth it to induce breaking changes?
1. Since [every single widget function gets wrapped every single time there is a page rerun](https://github.com/444B/streamlit-analytics2/discussions/56), we want to reduce this large performance overhead and have some functions run only when needed. For this, we will make use of [st.experimental_fragment](https://docs.streamlit.io/develop/api-reference/execution-flow/st.fragment), released in streamlit 1.33 this week
2. Since streamlit 1.33, there is now an [st.html](https://docs.streamlit.io/develop/api-reference/utilities/st.html). Currently, this is not supporting JS scripts but it may in the future. In either case, we want to create ways to get your analytics ported to GA4, and any other alternatives
3. Currently the password if used, is set via a function argument. We want to provide greater personalization and management for your app via a streamlit-analytics2-config.yml that hold your future settings for streamlit-analytics2
4. We want to support the tracking of multi page apps, which has been native to streamlit for quite some time and page links since 1.31
5. The app code does require a small refactor to separate UI, tracking and display logic. We also want to create further app test coverage with AppTest from streamlit (which should occur before a 1.0.0 release so we can ensure minimal breakage

Alot of these changes require some code rewrites and new things added, this we are aiming for a major change with hopefully lots of improvements in performance, maintenance, privacy and security

**What we want from you:**
Please share below any features or significant issues that are important to you or you wish to see

TLDR;
Feedback requested! I want to understand what you need, so that I can try provide it to you :)
Thanks!

**Comment by jgieseler** [comment]
Thanks for actively working on this package! For me important would be to have a way to persitently store the statistics, so that they don't get lost when resetting the app. For that, I'm only aware of the Firestore database integration, which would need https://github.com/444B/streamlit-analytics2/issues/21 to be fixed. Afaik, that was more or less completed, but just hasn't been merged yet.

**Reply by 444B** [reply]
Thank you so much for the feedback @jgieseler 
Feedback is arguably one of the most important aspects of open source :)

I'll take a look into #21 and the upstream PR (which didn't get pulled by my migration script) and see if we can get that code tested

Will do so later today but feel free to make that PR from the associated branch

**Reply by 444B** [reply]
@jgieseler 
I had tagged you for visibility in #77 
This has been merged into the latest release of SA2, which is now version 0.7.2

Usage is as such:
1. add a .streamlit/secrets.toml in your root dir of your project
2. Add keys and values for the secrets.toml or just pass them into the function arguments as strings (or environment variables). Here is an example of a min. required secrets.toml
```toml
firebase_key = "firebase key, formatted as per python script below"
project_name = "streamlit-analytics2-1a0ac"
collection_secret = "streamlit-analytics2"
```
3. Run this python script to get the right formatting for your firebase_key value
```python
import toml
import os

# Create streamlit secrets directory and secrets.toml if it doesn't exist
secrets_dir = "./.streamlit"
secrets_file = f"{secrets_dir}/secrets.toml"

if not os.path.exists(secrets_dir):
    os.mkdir(secrets_dir)
    open(secrets_file, "a").close()  # 'a' mode to create the file if it doesn't exist

# Correctly specify the path to the JSON file
json_file_path = "path/to/firebase-key.json"  # Correctly specify the JSON filename as a string

# Read the JSON file content
with open(json_file_path) as json_file:
    json_text = json_file.read()

# Prepare the TOML configuration
config = {"firebase_key": json_text}
toml_config = toml.dumps(config)

# Write the TOML configuration to the secrets file
with open(secrets_file, "w") as target:
    target.write(toml_config)

```
4. Use it in your app
```python
with streamlit_analytics.track(verbose=True, streamlit_secrets_firestore_key="firebase_key", firestore_project_name=st.secrets["project_name"], firestore_collection_name=st.secrets["collection_secret"]):
```

You will need to know the collection name and the firebase project

I am aware that this feature takes a bit of toil and what I consider extra steps
I promise that in time with the 1.0.0 release and the introduction of a config screen, you will not need to configure SA2 via function arguments again.

I hope that is helpful! Ill add documentation to the wiki in the next few days to details the usage better
Kindly test it out and let me know of any feedback :)

**Reply by jgieseler** [reply]
Thanks a lot for working on this! 🙏 Unfortunately, I'm right now quite busy with deadlines, so I can't test is straight ahead.

**Reply by 444B** [reply]
@jgieseler of course!
It's been tested and to my memory works as expected 
Take your time :) gl

**Comment by emigre459** [comment]
RE: `st.experimental_fragment` (which I believe is now more fully supported via `st.fragment`), this is a mechanism we have tried to use many times in different forms to avoid app-wide reruns and often we have to rollback the implementation of it and remove it from the codebase. I admit I'm having trouble finding the example GH issues I usually link to when advising caution in usage of `st.fragment` (maybe latest version has fixed a lot of things?) but we have observed some weird behavior with it before especially when trying to implement javascript elements within streamlit (e.g. via the `streamlit-js-eval` package). So just wanted to flag.
